src/code_transformer.ts

```diff
@@ -0,0 +1,203 @@
+/*
+ * @adonisjs/assembler
+ *
+ * (c) AdonisJS
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+import { join } from 'node:path'
+import { fileURLToPath } from 'node:url'
+import type { CodeBlockWriter, Project, SourceFile } from 'ts-morph'
+
+import { AddMiddlewareEntry, EnvValidationDefinition } from './types.js'
+
+/**
+ * This class is responsible for updating
+ */
+export class CodeTransformer {
+  #cwd: URL
+
+  /**
+   * Reference to TS Morph
+   * We lazy load it to avoid loading it when not
+   * used
+   */
+  #tsMorph!: typeof import('ts-morph')
+
+  /**
+   * The TsMorph project
+   */
+  #project!: Project
+
+  constructor(cwd: URL) {
+    this.#cwd = cwd
+  }
+
+  /**
+   * Lazy load ts-morph and create the project
+   */
+  async #loadProject() {
+    if (this.#project) return
+
+    this.#tsMorph = await import('ts-morph')
+    this.#project = new this.#tsMorph.Project({
+      tsConfigFilePath: join(fileURLToPath(this.#cwd), 'tsconfig.json'),
+    })
+  }
+
+  /**
+   * Add a new middleware to the middleware array of the
+   * given file
+   */
+  #addToMiddlewareArray(file: SourceFile, target: string, middlewareEntry: AddMiddlewareEntry) {
+    const callExpressions = file
+      .getDescendantsOfKind(this.#tsMorph.SyntaxKind.CallExpression)
+      .filter((statement) => statement.getExpression().getText() === target)
+
+    if (!callExpressions.length) {
+      throw new Error(`Cannot find ${target} statement in the file.`)
+    }
+
+    const arrayLiteralExpression = callExpressions[0].getArguments()[0]
+    if (
+      !arrayLiteralExpression ||
+      !this.#tsMorph.Node.isArrayLiteralExpression(arrayLiteralExpression)
```

Comment:
not lazy-loading ts-morph and having a dedicated entry point would avoid this kind of annoying thing
but that's not a big deal either